airflow/io/xcom.py

```diff
@@ -0,0 +1,111 @@
+# under the License.
+from __future__ import annotations
+
+import json
+import sys
+import uuid
+from typing import TYPE_CHECKING, Any, TypeVar
+from urllib.parse import urlsplit
+
+from airflow.configuration import conf
+from airflow.io.path import ObjectStoragePath
+from airflow.models.xcom import BaseXCom
+from airflow.utils.json import XComDecoder, XComEncoder
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+    from airflow.models import XCom
+
+T = TypeVar("T")
+
+
+class XComObjectStoreBackend(BaseXCom):
+    """XCom backend that stores data in an object store."""
+
+    path = conf.get("core", "xcom_objectstore_path", fallback="")
+
+    @staticmethod
+    def _get_key(data: str) -> str:
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        p = ObjectStoragePath(path)
+
+        url = urlsplit(data)
+        if url.scheme:
+            k = ObjectStoragePath(data)
+
+            if k.is_relative_to(p) is False:
+                raise ValueError(f"Invalid key: {data}")
+            else:
+                return data.replace(path, "", 1).lstrip("/")
+
+        raise ValueError(f"Not a valid url: {data}")
+
+    @staticmethod
+    def serialize_value(
+        value: T,
+        *,
+        key: str | None = None,
+        task_id: str | None = None,
+        dag_id: str | None = None,
+        run_id: str | None = None,
+        map_index: int | None = None,
+    ) -> bytes | str:
+        s_val = json.dumps(value, cls=XComEncoder).encode("utf-8")
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        compression = conf.get("core", "xcom_objectstore_compression", fallback=None)
+        threshold = conf.getint("core", "xcom_objectstore_threshold", fallback=-1)
+
+        if path and -1 < threshold < sys.getsizeof(value):
+            p = ObjectStoragePath(path) / f"{run_id}/{task_id}/{str(uuid.uuid4())}"
+
+            if not p.exists():
+                p.parent.mkdir(parents=True, exist_ok=True)
+
+            with p.open("wb", compression=compression) as f:
+                f.write(s_val)
+
+            return BaseXCom.serialize_value(str(p))
+        else:
+            return s_val
+
+    @staticmethod
+    def deserialize_value(
+        result: XCom,
+    ) -> Any:
+        data = BaseXCom.deserialize_value(result)
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        try:
+            p = ObjectStoragePath(path) / XComObjectStoreBackend._get_key(data)
+            return json.load(p.open("rb"), cls=XComDecoder)
+        except TypeError:
+            return data
+        except ValueError:
+            return data
+
+    @staticmethod
+    def purge(xcom: XCom, session: Session) -> None:
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        if isinstance(xcom.value, str):
+            try:
+                p = ObjectStoragePath(path) / XComObjectStoreBackend._get_key(xcom.value)
+                p.unlink()
```

Comment:
I think we should have `missing_ok=True` here - this way  we handle purge also for low-threshold value here.

Reply:
Or just a case that somene else deleted it already (if the above is handled by the if above)

Reply:
good point